resolver/src/main/java/io/netty/resolver/HostsFileEntriesProvider.java

```diff
@@ -0,0 +1,278 @@
+/*
+ * Copyright 2021 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.resolver;
+
+import io.netty.util.NetUtil;
+import io.netty.util.internal.PlatformDependent;
+import io.netty.util.internal.logging.InternalLogger;
+import io.netty.util.internal.logging.InternalLoggerFactory;
+
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.io.Reader;
+import java.net.Inet4Address;
+import java.net.InetAddress;
+import java.nio.charset.Charset;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.regex.Pattern;
+
+import static io.netty.util.internal.ObjectUtil.checkNotNull;
+
+/**
+ * A container of hosts file entries
+ */
+public final class HostsFileEntriesProvider {
+
+    public interface Parser {
+
+        /**
+         * Specifies {@link Charset}s that are used one after another until the parser is able to return a result
+         * or none is left.
+         * If charsets is not specified, the system default {@link Charset} is used for decoding.
+         *
+         * @param charsets the {@link Charset}s to try as file encodings when parsing
+         * @return {@code this}
+         */
+        Parser charsets(Charset... charsets);
+
+        /**
+         * Specifies the hosts file to be parsed.
+         * If a hosts file is not specified, the hosts file at standard OS location is used.
+         *
+         * @param file the hosts file
+         * @return {@code this}
+         */
+        Parser file(File file);
```

Comment:
@violetagg I think the two methods above should not exists but this should be just arguments to `parse`

Reply:
So instead of configuration I should add the same signature as in `HostsFileParser`, is that correct?

```
parseSilently()
parseSilently(Charset... charsets)
parse()
parse(File file)
parse(File file, Charset... charsets)
```

Reply:
yes... Or alternative just one which take all parameter and allow "null".

Reply:
Please check the new commit